src/shared/cloudformation/templateRegistry.ts

```diff
@@ -0,0 +1,143 @@
+/*!
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import * as path from 'path'
+import { fileURLToPath, pathToFileURL } from 'url'
+import * as vscode from 'vscode'
+import { FileWatcherListener } from '../utilities/fileSystemWatcher'
+import { CloudFormation } from './cloudformation'
+
+let templateRegistry: CloudFormationTemplateRegistry | undefined
+
+export function getTemplateRegistry(): CloudFormationTemplateRegistry {
+    if (!templateRegistry) {
+        throw new Error(
+            'Template Registry not initialized. Extension code should call activate() from shared/cloudformation/activation, test code should call setTemplateRegistry().'
+        )
+    }
+
+    return templateRegistry
+}
+
+export function setTemplateRegistry(registry: CloudFormationTemplateRegistry | undefined): void {
+    templateRegistry = registry
+}
+
+export interface CloudFormationTemplateRegistry {
+    registeredTemplates: Map<string, CloudFormation.Template>
+    getRegisteredTemplate(templatePath: string): CloudFormation.Template | undefined
+    addTemplateToTemplateData(templatePath: vscode.Uri): Promise<void>
+    removeTemplateFromRegistry(templatePath: vscode.Uri): void
+}
+
+export class DefaultCloudFormationTemplateRegistryListener implements FileWatcherListener {
```

Comment:
why is `FileWatcherListener` and `DefaultCloudFormationTemplateRegistryListener` needed? could delete both of them without any loss of clarity.

Reply:
`FileWatcherListener` could feasibly be elsewhere. Specifically, this is a bridge from the event handlers for the `vscode.FileSystemWatcher` to any other chunk of functionality (in this case, the registry), which is why it's an extendable interface.

The other option is to forego the use of this listener, have all `vscode.FileSystemWatcher` objects instead emit an event when they receive one, and subscribe any "listeners" to those emitted events. I was debating that, but this seemed a bit more direct. Interested to hear your opinions.

Reply:
Listener and Watcher terminology is confusing from a maintenance perspective. 

This class is responsible for watching for events, and updating a registry as a result. Something like `CloudFormationTemplateRegistryManager` might be more appropriate.